### ISSUE TITLE
iOS: After several minutes of use, error_listen_failed errors may occur

### ISSUE THREAD

**Issue by kakobayashi** [csdcorp/speech_to_text#481]
Thank you for providing us with a great package.

I have confirmed with Crashlytics that the `error_listen_failed` error may occur.

- It occurs about 5~10 minutes after use.
- This does not occur on all devices. It is found to occur on about 5% of all devices.
- On devices where this occurs, this error occurs repeatedly.
- This error appears to be a channel count of 0.
    - https://github.com/csdcorp/speech_to_text/blob/main/speech_to_text/ios/Classes/SwiftSpeechToTextPlugin.swift#L404-L406
    ```swift
        if(inputNode?.inputFormat(forBus: 0).channelCount == 0){
            throw SpeechToTextError.runtimeError("Not enough available inputs.")
        }
    ```

### Device information
- Device: iPhone14, iPhone14Pro, iPhoneSE(3rd), iPhone11ProMax, iPhone13 Mini, iPhone11, iPhone12, iPad Air(4th), iPhone15, iPhone15 Pro ...
- iOS: 17.2.1, 17.1.1
- iPadOS: 17.2.0

### Flutter version
- 3.7.11, 3.16.7

I'm in trouble understanding the cause of this error.
I have forked this package, so I can log it into SwiftSpeechToTextPlugin.swift and provide the information.

Thank you.

**Comment by sowens-csd** [comment]
Thanks for the very thorough problem report. That's an interesting issue. I wonder if I'm leaking channels. 

1. Does the application use any other sound input, like WebRTC, sound recorder, etc? 
2. Do you have any idea how many sound recognition sessions did/could happen during the 10-15 minute period? 
3. Can you reproduce the problem?

**Comment by kakobayashi** [comment]
1. My application includes the [record](https://pub.dev/packages/record) package.

2. I will add the logs in my app and check it. I am also asking the user who is getting this error.

3. I have tried my iPhone15 Pro and could not get it to occur.
However, I can confirm that it is occurring on the iPhone15 Pro on Crashlytics.
I don't think it is a device issue.

---
Here is what I have tried to do

1. The error did not change when AudioSession was regenerated when channelCount was 0, as shown below.
https://github.com/kakobayashi/speech_to_text/blob/v6.4.1/autoPunctuation/speech_to_text/ios/Classes/SwiftSpeechToTextPlugin.swift#L397

2. I have removed the "Not enough available inputs." error throw process so that the details of the error can be checked on the Flutter side, as shown below.
https://github.com/kakobayashi/speech_to_text/blob/v6.4.1/autoPunctuation_v4/speech_to_text/ios/Classes/SwiftSpeechToTextPlugin.swift#L451-L452

I have checked the following logs

```
SpeechRecognitionError msg: error_listen_failed, permanent: true, detail: Error Domain=Try Code=1 "(null)" UserInfo={exception=required condition is false: IsFormatSampleRateAndChannelCountValid(hwFormat)} 0 48000.0
```

The `audioSession.sampleRate` was confirmed only at 48000.
I have confirmed over 100 cases at 48000, but not occurring at other times.

I tried setting `sampleRate` to 48000 as shown below, but this did not fix this error.

```dart
      await _speechToText.listen(
        localeId: localeId,
        listenMode: ListenMode.dictation,
        onDevice: false,
        onResult: onResult,
        sampleRate: 48000,  // add
      );
```

**Comment by sowens-csd** [comment]
My first thought, since you're also using the record package in the app, is that the microphone is in use at the time STT is trying to install the tap. I don't know whether in your app it would be possible for the user to not enable or at least not use the record functionality and see if they can still reproduce the issue? If not perhaps you could get them on TestFlight and get them a version that allows for that test?

**Comment by kakobayashi** [comment]
I can ask the user who is experiencing this error to test the app without the record package for testing.
I will try to distribute the app to this user on TestFlight.

**Comment by kakobayashi** [comment]
I distributed the app on TestFlight with the Record package removed.
However, that app is getting the same error.
I would like to know if you have any other thoughts.

**Comment by kakobayashi** [comment]
This error was occurring when the interstitial ads were not disposed properly.
Thank you very much.

**Comment by davidd647** [comment]
@kakobayashi Hi! Just wondering how you narrowed down the issue.

I'm having this problem, and I don't know what's not being disposed properly...

Chipping away at the issue bit by bit, but any information you could provide might be very useful!

**Comment by kakobayashi** [comment]
@davidd647 
Hi there.
Narrowing down this issue was quite challenging.
I disabled all the features that used the device's microphone or speakers and confirmed that the issue no longer occurred.
After that, I found out that the interstitial ads were the cause and discovered that they were not being disposed of properly.
If possible, I recommend you try the same approach to see if the problem stops occurring for you as well.
